### ISSUE TITLE
Datasets not reshuffling between epochs in eager mode

### ISSUE THREAD

**Issue by malmaud** [tensorflow/tensorflow#26179]
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): 10.14
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 1.13.0-dev20190227
- Python version: 3.7

**Describe the current behavior**
Iterating a shuffled dataset returns elements in the same order each time it is iterated over, and the order is unaffected by the random seed.

**Describe the expected behavior**
The order should be deterministic given the random seed, and there should be a way to reshuffle between epochs.

**Code to reproduce the issue**

```python
import tensorflow as tf

tf.enable_eager_execution()

d = tf.data.Dataset.range(5).shuffle(5)
tf.set_random_seed(0)
elems = [item.numpy() for item in d]
print("First epoch: ", elems)
elems = [item.numpy() for item in d]
print("Second epoch: ", elems)
```

Running this twice produces:

First run:
```
First epoch:  [4, 2, 1, 3, 0]
Second epoch:  [4, 2, 1, 3, 0]
```

Second run:
```
First epoch:  [3, 1, 0, 2, 4]
Second epoch:  [3, 1, 0, 2, 4]
```

So even though the same random seed is used in both runs, the order of elements differs between them. And additionally, the order is the same between epochs, which you wouldn't necessarily want in your training loop.

**Comment by malmaud** [comment]
Actually, it looks like this is all explained by `d.shuffle` caching the random seed when it's called instead of when the dataset is iterated over. That seems a bit strange since it means I need to recall `shuffle` at the start of each epoch and so eg passing a dataset to Keras's `fit` method is going to use iteration over for every epoch.

**Comment by mrry** [comment]
Thanks for reporting this problem! This is definitely surprising behavior, and we should find a way to make it easier.

@rohan100jain Do you think it would be possible to extend your solution from b397935bae2530e1cf5022ca6f85a0b898a795d6 to work as expected in the eager case?

Also CCing @wangpengmit because whatever happens here ought to be compatible with [https://github.com/tensorflow/community/blob/master/rfcs/20181217-tf2-random-numbers.md](the TF 2.0 randomness interface), and @jsimsa FYI.

**Comment by rob-rowe** [comment]
This still occurs in tf2.0.0-beta1. In my case I have a multitask training scenario where the training set sizes are very different. I would like to set a Keras epoch to correspond to the smaller training set size but because of this issue (or something related to it) it appears that the training data for the task with the larger training set size resets and repeats. This effectively truncates the size of the large training set to be ~same as the small one. Is there a work-around other than setting the epoch to correspond to the large training set size?

**Comment by jsimsa** [comment]
A workaround for this problem is to set the `seed` argument of the shuffle transformation to 0. This will result in reshuffling each epoch. However, the behavior will not be deterministic. 

Being able to provide deterministic reshuffling each epoch requires maintaining the RNG state outside of the tf.data graph and is currently not supported.

**Comment by rob-rowe** [comment]
@jsimsa When I add seed=0 to shuffle() in the sample code given by @malmaud, the results don't appear to change: the random order is the same between epochs. Also it appears that 0 provides a deterministic order similar to other values.

**Comment by jsimsa** [comment]
@rob-rowe mea culpa, as it turns out the argument only controls the op-level seed and the graph-level seed will still end up being non-zero. I have updated my original answer.

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.

**Comment by sachinprasadhs** [comment]
@malmaud , you can mention seed inside shuffle to produce shuffled data between epochs. Please find the gist [here](https://colab.research.google.com/gist/sachinprasadhs/7285e92602be3455e8f60ca104c2b0f6/26179.ipynb) with the working code in Tensorflow 2.5.
